prow/cmd/pipeline/controller.go

```diff
@@ -0,0 +1,673 @@
+	cfg, ok := c.pipelines[ctx]
+	if !ok {
+		defaultCtx := kube.DefaultClusterAlias
+		defaultCfg, ok := c.pipelines[defaultCtx]
+		if !ok {
+			return pipelineConfig{}, fmt.Errorf("no cluster configuration found for default context %q", defaultCtx)
+		}
+		return defaultCfg, nil
+	}
+	return cfg, nil
+}
+
+func (c *controller) getProwJob(name string) (*prowjobv1.ProwJob, error) {
+	return c.pjLister.ProwJobs(c.pjNamespace()).Get(name)
+}
+
+func (c *controller) updateProwJob(pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error) {
+	logrus.Debugf("updateProwJob(%s)", pj.Name)
+	return c.pjc.ProwV1().ProwJobs(c.pjNamespace()).Update(pj)
+}
+
+func (c *controller) getPipelineRun(context, namespace, name string) (*pipelinev1alpha1.PipelineRun, error) {
+	p, err := c.getPipelineConfig(context)
+	if err != nil {
+		return nil, err
+	}
+	return p.informer.Lister().PipelineRuns(namespace).Get(name)
+}
+
+func (c *controller) getPipelineRunWithSelector(context, namespace, selector string) (*pipelinev1alpha1.PipelineRun, error) {
+	p, err := c.getPipelineConfig(context)
+	if err != nil {
+		return nil, err
+	}
+
+	label, err := labels.Parse(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse selector %s", selector)
+	}
+	runs, err := p.informer.Lister().PipelineRuns(namespace).List(label)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list pipelineruns with label %s", label.String())
+	}
+	if len(runs) > 1 {
+		return nil, fmt.Errorf("%s pipelineruns found with label %s, expected only 1", string(len(runs)), label.String())
+	}
+	if len(runs) == 0 {
+		return nil, apierrors.NewNotFound(pipelinev1alpha1.Resource("pipelinerun"), label.String())
+	}
+	return runs[0], nil
+}
+
+func (c *controller) deletePipelineRun(context, namespace, name string) error {
+	logrus.Debugf("deletePipeline(%s,%s,%s)", context, namespace, name)
+	p, err := c.getPipelineConfig(context)
+	if err != nil {
+		return err
+	}
+	return p.client.TektonV1alpha1().PipelineRuns(namespace).Delete(name, &metav1.DeleteOptions{})
+}
+func (c *controller) createPipelineRun(context, namespace string, p *pipelinev1alpha1.PipelineRun) (*pipelinev1alpha1.PipelineRun, error) {
+	logrus.Debugf("createPipelineRun(%s,%s,%s)", context, namespace, p.Name)
+	pc, err := c.getPipelineConfig(context)
+	if err != nil {
+		return nil, err
+	}
+	return pc.client.TektonV1alpha1().PipelineRuns(namespace).Create(p)
+}
+
+func (c *controller) createPipelineResource(context, namespace string, pr *pipelinev1alpha1.PipelineResource) (*pipelinev1alpha1.PipelineResource, error) {
+	logrus.Debugf("createPipelineResource(%s,%s,%s)", context, namespace, pr.Name)
+	pc, err := c.getPipelineConfig(context)
+	if err != nil {
+		return nil, err
+	}
+	return pc.client.TektonV1alpha1().PipelineResources(namespace).Create(pr)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+func (c *controller) pipelineID(pj prowjobv1.ProwJob) (string, error) {
```

Comment:
This function doesn't seem right to me. In particular it will always fail for periodic jobs because periodics always have nil `Spec.Refs`.
Why was this changed from what the build controller does? https://github.com/kubernetes/test-infra/blob/b7b7a8907608cc78abae88860367eb6710157aa5/prow/cmd/build/controller.go#L345-L352

Reply:
I think, we wanted to generate build number per branch instead of prow job which for row pipeline scenario is not relevant. 

I will use the same logic like in the build controller to generate the build IDs.

Reply:
used the same way to generate the build ID like in the build controller